File: components/auth/src/test/java/com/ecfront/dew/auth/BasicTest.java
package com.ecfront.dew.auth;

import com.ecfront.dew.common.Page;
import com.ecfront.dew.common.Resp;
import com.ecfront.dew.core.Dew;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.web.client.TestRestTemplate;

import java.util.List;


public abstract class BasicTest {

    @Autowired
    private TestRestTemplate testRestTemplate;

    private String token;

    protected String u(String uri) {
        if (uri.contains("?")) {
            return uri + "&" + Dew.Constant.TOKEN_VIEW_FLAG + "=" + token;
        } else {
            return uri + "?" + Dew.Constant.TOKEN_VIEW_FLAG + "=" + token;
        }
    }

    protected <E> Resp<E> postAndReturnObj(String uri, Object body, Class<E> clazz) {
        return Resp.generic(testRestTemplate.postForObject(u(uri), body, Resp.class), clazz);
    }

    protected <E> Resp<List<E>> postAndReturnList(String uri, Object body, Class<E> clazz) {
        return Resp.genericList(testRestTemplate.postForObject(u(uri), body, Resp.class), clazz);
    }

    protected <E> Resp<Page<E>> postAndReturnPage(String uri, Object body, Class<E> clazz) {
        return Resp.genericPage(testRestTemplate.postForObject(u(uri), body, Resp.class), clazz);
    }

    protected void put(String uri, Object body) {
        testRestTemplate.put(u(uri), body, Resp.class);
    }

    protected <E> Resp<E> getAndReturnObj(String uri, Class<E> clazz) {
        return Resp.generic(testRestTemplate.getForObject(u(uri), Resp.class), clazz);
    }

    protected <E> Resp<List<E>> getAndReturnList(String uri, Class<E> clazz) {
        return Resp.genericList(testRestTemplate.getForObject(u(uri), Resp.class), clazz);
    }

    protected <E> Resp<Page<E>> getAndReturnPage(String uri, Class<E> clazz) {
        return Resp.genericPage(testRestTemplate.getForObject(u(uri), Resp.class), clazz);
    }

    protected void delete(String uri) {
        testRestTemplate.delete(u(uri), Resp.class);
    }

}

File: components/auth/src/test/scala/com/tairanchina/csp/dew/auth/BasicTest.scala
package com.tairanchina.csp.dew.auth

import java.util

import com.ecfront.dew.common.{Page, Resp}
import com.tairanchina.csp.dew.auth.sdk.AuthSDKConfig
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.http.{HttpEntity, HttpHeaders, HttpMethod}


abstract class BasicTest {

  @Autowired
  private var testRestTemplate: TestRestTemplate = _

  protected var accessToken: String = ""
  protected var token: String = ""

  protected def post[E](uri: String, body: Any, clazz: Class[E]): Resp[E] = {
    Resp.generic(exchange(HttpMethod.POST, uri, body), clazz)
  }

  protected def postList[E](uri: String, body: Any, clazz: Class[E]): Resp[util.List[E]] = {
    Resp.genericList(exchange(HttpMethod.POST, uri, body), clazz)
  }

  protected def postPage[E](uri: String, body: Any, clazz: Class[E]): Resp[Page[E]] = {
    Resp.genericPage(exchange(HttpMethod.POST, uri, body), clazz)
  }

  protected def put[E](uri: String, body: Any, clazz: Class[E]): Resp[E] = {
    Resp.generic(exchange(HttpMethod.PUT, uri, body), clazz)
  }

  protected def putList[E](uri: String, body: Any, clazz: Class[E]): Resp[util.List[E]] = {
    Resp.genericList(exchange(HttpMethod.PUT, uri, body), clazz)
  }

  protected def putPage[E](uri: String, body: Any, clazz: Class[E]): Resp[Page[E]] = {
    Resp.genericPage(exchange(HttpMethod.PUT, uri, body), clazz)
  }

  protected def get[E](uri: String, clazz: Class[E]): Resp[E] = {
    Resp.generic(exchange(HttpMethod.GET, uri, null), clazz)
  }

  protected def getList[E](uri: String, clazz: Class[E]): Resp[util.List[E]] = {
    Resp.genericList(exchange(HttpMethod.GET, uri, null), clazz)

  }

  protected def getPage[E](uri: String, clazz: Class[E]): Resp[Page[E]] = {
    Resp.genericPage(exchange(HttpMethod.GET, uri, null), clazz)
  }

  protected def delete(uri: String): Unit = {
    exchange(HttpMethod.DELETE, uri, null)
  }

  private def exchange[E](method: HttpMethod, uri: String, body: Any): Resp[_] = {
    val requestHeaders = new HttpHeaders
    requestHeaders.add(AuthSDKConfig.HTTP_ACCESS_TOKEN, accessToken)
    requestHeaders.add(AuthSDKConfig.HTTP_USER_TOKEN, token)
    val requestEntity = new HttpEntity[Any](body, requestHeaders)
    testRestTemplate.exchange(uri, method, requestEntity, classOf[Resp[_]]).getBody
  }

}
